### PR TITLE
[ML] Improve normalisation of anomaly detection results for short bucket lengths

### LINKED ISSUE #2276: [ML] Revisit correction for bucket length when computing probability quantiles normalising results
Currently, we try and account for the different number of results per time interval for different bucket lengths when computing probability quantiles as part of rate limiting normalised scores.

The assumption we make should result in roughly the same score distribution when modelling pure noise (I need to recheck this). However, it will tend to pull down scores when there are long lasting anomalies in the data for short bucket lengths. This has probably been exacerbated by multi-bucket anomaly detection. In particular, we see cases that the correction has the reverse impact to the one we'd like: producing less consistent scoring for different bucket lengths.

We should revisit whether any sort of bucket length correction rate limiting normalised scores is a good idea.

### PR DESCRIPTION
**Pull request by tveasey**
This changes the way we normalise for short bucket lengths.

Previously, we tried to apply consistent rate limiting of alerts across different bucket lengths by accounting for the different number of results we receive per time interval. The approach we took proved to interact badly with uncertainty in the estimates of extreme quantiles. The upshot is we actually could get bias where our scores would be too low short bucket lengths. The problem was particularly severe if the time series had extended runs of unusual values.

This takes a different approach: we simply only update the quantiles with the high water mark in each fixed time interval. This achieves the same thing, but works reliably for shorter bucket lengths. I've added a test that rate limiting is working as expected across multiple bucket lengths.

This will potentially cause a non-trivial results change. One thing to consider is if we need special handling for upgrading existing jobs. First off we need to see differences on our QA suite.

Fixes #2276.

### CODE REVIEW

File: include/model/CAnomalyScore.h [hunk 1]
```diff
@@ -268,13 +268,13 @@ class MODEL_EXPORT CAnomalyScore {
...
         //! The maximum possible normalized score.
-        double m_MaximumNormalizedScore = 100.0;
+        double m_MaximumNormalizedScore{100.0};
 
         //! The approximate HIGH_PERCENTILE percentile raw score.
         uint32_t m_HighPercentileScore;
         //! The number of scores less than the approximate
         //! HIGH_PERCENTILE percentile raw score.
-        uint64_t m_HighPercentileCount = 0;
+        uint64_t m_HighPercentileCount{0};
```
Comment (droberts195) [inline include/model/CAnomalyScore.h:277]:
```suggestion
        std::uint64_t m_HighPercentileCount{0};
```

File: include/model/CAnomalyScore.h [hunk 2]
```diff
@@ -268,13 +268,13 @@ class MODEL_EXPORT CAnomalyScore {
         //! The normalized anomaly score knot points.
         TDoubleDoublePrVec m_NormalizedScoreKnotPoints;
         //! The maximum possible normalized score.
-        double m_MaximumNormalizedScore = 100.0;
+        double m_MaximumNormalizedScore{100.0};
 
         //! The approximate HIGH_PERCENTILE percentile raw score.
         uint32_t m_HighPercentileScore;
```
Comment (droberts195) [inline include/model/CAnomalyScore.h:274]:
```suggestion
        std::uint32_t m_HighPercentileScore;
```

File: lib/api/unittest/CResultNormalizerTest.cc [hunk 3]
```diff
@@ -60,7 +60,7 @@ BOOST_AUTO_TEST_CASE(testInitNormalizerPartitioned) {
 
     // The maximum bucket_time influencer probability in the farequote data used to initialise
     // the normaliser is 2.1305076675260463E-24, so this should map to the highest normalised
-    // score which is 90.69091
+    // score which is 90.69091.
```
Comment (droberts195) [inline lib/api/unittest/CResultNormalizerTest.cc:63]: This comment no longer matches the value on line 81.

File: lib/model/CAnomalyScore.cc [hunk 4]
```diff
@@ -271,23 +279,19 @@ bool CAnomalyScore::CNormalizer::canNormalize() const {
...
         // Nothing to do.
         return true;
     }
-
     if (m_RawScoreQuantileSummary.n() == 0) {
-        LOG_ERROR(<< "No scores have been added to the quantile summary");
-        return false;
+        score = 0.0;
+        return true;
     }
```
Comment (droberts195) [inline lib/model/CAnomalyScore.cc:289]: Are we intentionally going through this code branch now?  I would have thought it's a coding error to ask to normalize a score when we've never generated any score.

Comment (tveasey) [reply lib/model/CAnomalyScore.cc:289]: We don't update the quantiles with every score any more for short bucket lengths. So at the very beginning we hit the situation we haven't yet added any scores. I think this is harmless: it is only the first 30 mins of modelling, in practice scores always work out as zero at this point anyway. I don't think it is worthwhile trying to detect that we're not in start up, so I just don't bother to error in this case anymore.

File: include/model/CAnomalyScore.h (hunk 2 above)
Comment (tveasey) [reply include/model/CAnomalyScore.h:274]: My plan is to try and address some fo the clang tidy suggestions we're getting now we run it over the codebase. As part of this work I'm going to do one big change to switch all our uses of `stdint.h` to `cstdint` and migrate to the std namespace versions. I think I can do this with regex and I reckon it is better than making piecemeal changes for this.

File: include/model/CAnomalyScore.h (hunk 1 above)
Comment (tveasey) [reply include/model/CAnomalyScore.h:277]: See below.

**Review by droberts195 — APPROVED** [review]
LGTM

**Pull request merged**